File: src/dtos/crawler.rs
use std::collections::HashSet;

use educe::Educe;
use indexmap::IndexSet;
use serde::{Deserialize, Serialize};
use ts_rs::TS;

use crate::dtos::MediaType;

/// Template Variables:
/// - {query} - title of anime/manga + episode/chapter of anime/manga
#[derive(Debug, Deserialize, Educe)]
#[educe(Default)]
#[serde(default)]
pub struct CrawlerConfig {
    pub id: String,
    pub name: String,
    pub base_url: String,
    pub item_selector: String,
    pub title_selector: String,
    pub link_selector: String,
    pub popularity_selector: Option<String>,
    pub size_selector: Option<String>,

    #[educe(Default = true)]
    pub is_active: bool,

    #[educe(Default = HashSet::from([MediaType::Anime]))]
    pub category: HashSet<MediaType>,
}

impl CrawlerConfig {
    /// Returns a default crawler configuration for Nyaa.si anime torrents.
    /// Last updated on 23-Aug-2026
    pub fn fallback() -> Self {
        Self {
            id: "nyaa".into(),
            name: "Nyaa.si Anime Torrents".into(),
            base_url: "https://nyaa.si/?f=0&c=1_2&q={title}+{number}".into(),
            item_selector: "table.torrent-list tbody tr".into(),
            title_selector: "table.torrent-list tbody tr".into(),
            link_selector: "a[href^='/view/']:not(.comments)".into(),
            popularity_selector: Some("td:nth-child(6)".into()),
            size_selector: Some("td:nth-child(4)".into()),
            ..Default::default()
        }
    }
}

#[derive(Serialize, Deserialize, Default, TS)]
#[serde(default)]
#[ts(export)]
pub struct CrawlerResult {
    pub title: String,
    pub link: String,
    pub source: String,
    pub popularity: Option<String>,
    pub size: Option<String>,
    pub parsed_title: ParsedTitle,
    pub category: MediaType,
}

macro_rules! make_parsed_title {
    ($($field:ident),* $(,)?) => {
        #[derive(Debug, Clone, Serialize, Deserialize, Default, TS, PartialEq)]
        #[ts(export)]
        pub struct ParsedTitle {
            $(
                #[serde(skip_serializing_if = "IndexSet::is_empty")]
                #[ts(type = "string[] | undefined")]
                pub $field: IndexSet<String>,
            )*

            // Manually add `kind` since it needs the extra `rename` attribute
            #[serde(rename = "type", skip_serializing_if = "IndexSet::is_empty")]
            #[ts(type = "string[] | undefined")]
            pub kind: IndexSet<String>,
        }
    };
}

// Just list your fields here. The macro does the rest!
make_parsed_title!(
    audio_term,
    device,
    episode,
    episode_title,
    file_checksum,
    file_extension,
    language,
    other,
    part,
    release_group,
    release_information,
    release_version,
    season,
    source,
    subtitles,
    title,
    video_resolution,
    video_term,
    volume,
    year,
    episode_alt,
    date
);

File: src/dto/crawler.go
package dto

import (
	"encoding/json"
	"net/url"
	"reflect"
	"strings"

	"github.com/Oudwins/zog"
	mapset "github.com/deckarep/golang-set/v3"
)

type CrawlerResult struct {
	Title      string    `json:"title"`
	Link       string    `json:"link"`
	Source     string    `json:"source"`
	Popularity *string   `json:"popularity"`
	Size       *string   `json:"size"`
	Category   MediaType `json:"category"`

	ParsedTitle *ParsedTitle `json:"parsed_title"`
}

var ICrawlerResult = zog.Struct(zog.Shape{
	"Title": zog.String().Min(1),
	"Link": zog.String().TestFunc(func(v *string, ctx zog.Ctx) bool {
		u, err := url.Parse(*v)
		return err == nil && u.Scheme != ""
	}),
})

type ParsedTitle struct {
	AudioTerm          mapset.Set[string] `json:"audio_term,omitempty"`
	Device             mapset.Set[string] `json:"device,omitempty"`
	Episode            mapset.Set[string] `json:"episode,omitempty"`
	EpisodeTitle       mapset.Set[string] `json:"episode_title,omitempty"`
	FileChecksum       mapset.Set[string] `json:"file_checksum,omitempty"`
	FileExtension      mapset.Set[string] `json:"file_extension,omitempty"`
	Language           mapset.Set[string] `json:"language,omitempty"`
	Other              mapset.Set[string] `json:"other,omitempty"`
	Part               mapset.Set[string] `json:"part,omitempty"`
	ReleaseGroup       mapset.Set[string] `json:"release_group,omitempty"`
	ReleaseInformation mapset.Set[string] `json:"release_information,omitempty"`
	ReleaseVersion     mapset.Set[string] `json:"release_version,omitempty"`
	Season             mapset.Set[string] `json:"season,omitempty"`
	Source             mapset.Set[string] `json:"source,omitempty"`
	Subtitles          mapset.Set[string] `json:"subtitles,omitempty"`
	Title              mapset.Set[string] `json:"title,omitempty"`
	VideoResolution    mapset.Set[string] `json:"video_resolution,omitempty"`
	VideoTerm          mapset.Set[string] `json:"video_term,omitempty"`
	Volume             mapset.Set[string] `json:"volume,omitempty"`
	Year               mapset.Set[string] `json:"year,omitempty"`
	EpisodeAlt         mapset.Set[string] `json:"episode_alt,omitempty"`
	Date               mapset.Set[string] `json:"date,omitempty"`
	Kind               mapset.Set[string] `json:"kind,omitempty"`
	Unknown            mapset.Set[string] `json:"unknown,omitempty"`
}

func (p *ParsedTitle) UnmarshalJSON(data []byte) error {
	var aux map[string][]string
	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}

	val := reflect.ValueOf(p).Elem()
	typ := val.Type()

	for i := 0; i < typ.NumField(); i++ {
		field := typ.Field(i)
		tag := field.Tag.Get("json")
		jsonKey, _, _ := strings.Cut(tag, ",")

		if jsonKey == "" {
			jsonKey = field.Name
		}

		if slice, ok := aux[jsonKey]; ok && slice != nil {
			val.Field(i).Set(reflect.ValueOf(mapset.NewSet(slice...)))
		}
	}
	return nil
}
